src/ir/module-splitting.cpp

```diff
@@ -90,81 +88,118 @@ namespace {
-  for (auto& segment : table.segments) {
-    assert(segment.offset->is<Const>() &&
-           "TODO: handle non-constant segment offsets");
-    Index segmentBase = segment.offset->cast<Const>()->value.geti32();
-    if (segmentBase + segment.data.size() >= maxIndex) {
-      maxIndex = segmentBase + segment.data.size();
-      activeSegment = &segment;
-      activeBase = segmentBase;
+  // If there is exactly one table segment and that segment has a non-constant
+  // offset, append new items to the end of that segment. In all other cases,
+  // append new items at constant offsets after all existing items at constant
+  // offsets.
+  if (table.segments.size() == 1 && !table.segments[0].offset->is<Const>()) {
+    assert(table.segments[0].offset->is<GlobalGet>() &&
+           "Unexpected initializer instruction");
+    activeSegment = &table.segments[0];
+    activeBase = {table.segments[0].offset->cast<GlobalGet>()->name, 0};
+  } else {
+    // Finds the segment with the highest occupied table slot so that new items
+    // can be inserted contiguously at the end of it without accidentally
+    // overwriting any other items. TODO: be more clever about filling gaps in
+    // the table, if that is ever useful.
+    Index maxIndex = 0;
+    for (auto& segment : table.segments) {
+      assert(segment.offset->is<Const>() &&
+             "Unexpected non-const segment offset");
+      Index segmentBase = segment.offset->cast<Const>()->value.geti32();
+      if (segmentBase + segment.data.size() >= maxIndex) {
+        maxIndex = segmentBase + segment.data.size();
+        activeSegment = &segment;
+        activeBase = {"", segmentBase};
+      }
     }
   }
-
   // Initialize funcIndices with the functions already in the table.
-  forEachElement(table, [&](Index index, Name func) { addIndex(func, index); });
+  forEachElement(table, [&](Name base, Index offset, Name func) {
+    addSlot(func, {base, offset});
+  });
 }
 
-Index TableSlotManager::getIndex(Name func) {
-  auto indexIt = funcIndices.find(func);
-  if (indexIt != funcIndices.end()) {
-    return indexIt->second;
+auto TableSlotManager::getSlot(Name func) -> Slot {
+  auto slotIt = funcIndices.find(func);
+  if (slotIt != funcIndices.end()) {
+    return slotIt->second;
   }
 
   // If there are no segments yet, allocate one.
   if (activeSegment == nullptr) {
     table.exists = true;
     assert(table.segments.size() == 0);
     table.segments.emplace_back(Builder(module).makeConst(int32_t(0)));
-    activeSegment = &table.segments.front();
+    activeSegment = &table.segments.back();
   }
 
-  Index newIndex = activeBase + activeSegment->data.size();
+  Slot newSlot = {activeBase.global,
+                  activeBase.index + Index(activeSegment->data.size())};
   activeSegment->data.push_back(func);
-  addIndex(func, newIndex);
-  if (table.initial <= newIndex) {
-    table.initial = newIndex + 1;
+  addSlot(func, newSlot);
+  if (table.initial <= newSlot.index) {
```

Comment:
Right, there's not really anything to do for an unknown global base. Keeping this code path means that at least setting the base to 0 will work. I don't know if that's useful in practice, but it doesn't seem like it would hurt anything either.